### PR TITLE
Fix bug where submitting a reason would panic

### PR DESCRIPTION
**Pull request by euank**
Forgot to check reason length before slicing it. I added a test that catches this and made sure it passes.

Mind reviewing? @samuelkarp or @jhspaybar

### CODE REVIEW

File: agent/api/api_client.go [hunk 1]
```diff
@@ -77,8 +78,20 @@ func (client *ApiECSClient) serviceClient() (*svc.AmazonEC2ContainerServiceV2014
 	return ecs, nil
 }
 
+// serviceClient provides a client for interacting with the ECS service apis
+func (client *ApiECSClient) serviceClient() (svc.AmazonEC2ContainerServiceV20141113, error) {
+	return client.serviceClientFn()
+}
+
 func NewECSClient(credentialProvider credentials.AWSCredentialProvider, config *config.Config, insecureSkipVerify bool) ECSClient {
```
Comment (samuelkarp) [inline agent/api/api_client.go]: Minor: I'd probably reorder this so NewECSClient is first, serviceClient is second, and serviceClientImpl is third.

File: agent/api/api_client_test.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	svc "github.com/aws/amazon-ecs-agent/agent/ecs_client/ecs_autogenerated_client"
+)
+
+type MockAmazonEC2ContainerServiceV20141113Client struct {
+	svc.AmazonEC2ContainerServiceV20141113Client
+
+	lastRequest interface{}
+}
+
+type TestSubmitContainerStateChangeResponse struct {
```
Comment (samuelkarp) [inline agent/api/api_client_test.go]: This appears unused.

File: agent/api/api_client_test.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	return svc.NewSubmitContainerStateChangeResponse(), nil
+}
+
+func TestSubmitContainerStateChange(t *testing.T) {
+	client := NewECSClient(auth.TestCredentialProvider{}, &config.Config{Cluster: "default"}, false)
+	mockSvcClient := &MockAmazonEC2ContainerServiceV20141113Client{}
+	client.(*ApiECSClient).serviceClientFn = func() (svc.AmazonEC2ContainerServiceV20141113, error) {
+		return mockSvcClient, nil
+	}
+	err := client.SubmitContainerStateChange(ContainerStateChange{
```
Comment (samuelkarp) [inline agent/api/api_client_test.go]: I'd prefer that each of these SubmitContainerStateChange invocations was a separate test.

**Comment by euank** [comment]
@samuelkarp Thanks. I made all the changes you suggested.

File: agent/api/api_client_test.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	lastRequest interface{}
+}
+
+func (mock *MockAmazonEC2ContainerServiceV20141113Client) SubmitContainerStateChange(req svc.SubmitContainerStateChangeRequest) (svc.SubmitContainerStateChangeResponse, error) {
+	mock.lastRequest = req
+
+	return svc.NewSubmitContainerStateChangeResponse(), nil
+}
+
+func NewMockClient() (ECSClient, *MockAmazonEC2ContainerServiceV20141113Client) {
```
Comment (samuelkarp) [inline agent/api/api_client_test.go]: Minor: This does not need to be public.

**Comment by samuelkarp** [comment]
LGTM aside from the minor comment above.

File: agent/api/api_client_test.go (hunk 4 above)
Comment (euank) [inline agent/api/api_client_test.go]: I'm okay with it being public because it's in _test anyways, so nothing can rely on it.

**Pull request merged**